### DISCUSSION TITLE
Ternary comparison expressions 0 < x < 5

### DISCUSSION THREAD

**Discussion by JohanLarsson** [dotnet/csharplang#4106] [General]
Comparing a value with a lower and upper bound is common.
Writing `x > 0 && x < 5` includes duplication of the `x` identifier and hence a way to get it wrong.

**Comment by canton7** [comment]
Just to note, you can write:

```cs
x is (>0 and <5)
```

**Reply by 333fred** [reply]
Would be `and`, not `or`, but otherwise yes

**Reply by Xyncgas** [reply]
doesn't work with non constants

**Comment by HaloFour** [comment]
https://github.com/dotnet/roslyn/issues/136#issuecomment-239932932

**Reply by canton7** [reply]
E.g. [this diabolical code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYBGAWAChqBmAAgG9HXvXqqOuPIQDcAhlFYAKBGlYAvAJSsAvFLwwA7qwDCkhbPVbdChQG5BQ7hRoBOaawA8rAKyP5Zi6wC+jHw0YUbADOGLhgGDqMnAxCgbw0SKzAEBAANqwQAA4wUKIY0I6S2qwyOqwAnkrKAHysoTgw5jE8cdaJyWkZ2bn5EtVFJbLFlSq19Y2enq0JZVk5eQUOA6UAlngRIzUlTbFsbbPdC33LsmsbVbUITV5AA===)

**Reply by CyrusNajmabadi** [reply]
Yes.  ths is technically ambiguous.  However, i will champion this as i think we could say: if hte existing semantics are in error (which would be expected 99.999% of the time), then use these new semantics.

**Reply by HaloFour** [reply]
This would be nice, especially since you can't use non-constants in patterns ... (yet...)

**Reply by alrz** [reply]
> (yet...)

Yup. As you suggested elsewhere, it could be relaxed for readonly variables, and that makes this feature completely unnecessary.

**Reply by CyrusNajmabadi** [reply]
It would not make it unnecessary.  For example, `y is > y and < z` is extremely clunky and non-intuitive to read.  For most developers, `x < y < z` will be absolutely trivial to write and read/comprehend.  That's actually the primary reason i'm championing this.  Because i don't think general pattern forms are suitable here.  Yes, they can express hte same concept, but in a very non-idiomatic and clunky fashion.

**Reply by alrz** [reply]
To that end, can I suggest postfix relational patterns instead?

```cs
is y > and > z
```
There will be some nasty ambiguities but I think it's worth considering.

**Reply by CyrusNajmabadi** [reply]
> To that end, can I suggest postfix relational patterns instead?

Feel free (in a different discussion) :)

**Reply by DavidArno** [reply]
@CyrusNajmabadi,

Going to disagree with you on that one. I think `x is > y and < z` is highly intuitive as it uses existing pattern rules, just with variables rather than constants.

@alrz

Sorry but I really dislike that syntax.

**Reply by HaloFour** [reply]
@DavidArno 

> as it uses existing pattern rules

I'd imagine that a significantly smaller segment of the whole C# population is familiar with patterns, let alone relational patterns and their rules.

I think the ability to chain the comparison operators as suggested makes sense and reads nicely.  It is technically ambiguous but only in a very niche case that could remain supported without impacting the implementation.

I also think it's worthwhile opening patterns to non-constant expressions, but I think that's a much bigger conversation in general.

**Reply by alrz** [reply]
>I'd imagine that a significantly smaller segment of the whole C# population is familiar with patterns,

Isn't this also a *new feature*? So the same barrier applies only that patterns are actually a few releases ahead in terms of reach.

**Reply by HaloFour** [reply]
@alrz 

> Isn't this also a new feature?

It's not just a matter of which version of the language that you're on but also which features you're more likely to use.  I'd suggest that nearly all C# developers have used comparison operators where a significantly smaller percentage have used patterns in general, and a much smaller percentage have used relational operators.

**Reply by DavidArno** [reply]
@HaloFour,

You are likely right that only a small segment of the C# community have used relational patterns. My thoughts though were along these lines:

The team add the ability to express a range comparison as `y < x < z`
The team also add support for variables in patterns, so it also becomes possible to express it as `x is > y and < z` as a side effect

This will then lead to the usual chorus of "why are there two ways of doing the same thing?" So I'd suggest it best to ask the LDM the question "will you ever support variables in patterns?" If no, then the case for supporting `y < x < z` becomes strong. But if the answer is yes, then why create a second way of doing the same thing?

Of course I appreciate the answer will likely be "maybe" and this whole argument then falls down :)

**Comment by CyrusNajmabadi** [comment]
Now championed at https://github.com/dotnet/csharplang/issues/4108.  Will bring to a future triage to determine how LDM feels about this.

**Reply by bernd5** [reply]
For me that syntax looks strange - I would prefer to extend relational patterns to work with non - constants.
It looks like you would compare a boolean to something else.
